core/connection/tlsconfig.go

```diff
@@ -63,23 +75,52 @@ func (c *TLSConfig) AddClientCertFromFile(clientFile string) (string, error) {
 		return "", err
 	}
 
-	cert, err := tls.X509KeyPair(data, data)
+	var currentBlock *pem.Block
+	var certBlock, certDecodedBlock, keyBlock []byte
+
+	remaining := data
+	start := 0
+	for {
+		currentBlock, remaining = pem.Decode(remaining)
+		if currentBlock == nil {
+			break
+		}
+
+		if currentBlock.Type == "CERTIFICATE" {
+			certBlock = data[start : len(data)-len(remaining)]
+			certDecodedBlock = currentBlock.Bytes
+			start += len(certBlock)
+		} else if strings.HasSuffix(currentBlock.Type, "PRIVATE KEY") {
+			if len(certBlock) == 0 {
+				return "", fmt.Errorf("failed to find CERTIFICATE but did find private key; PEM inputs may be switched")
```

Comment:
I'd prefer not to require the PEM blocks to be in a particular order, as it's pretty trivial to flip it around for users and DWIM.  You can see an approach adapted from this file in the TOOLS project: https://github.com/mongodb/mongo-tools/blob/master/common/db/tlsgo/config.go

Reply:
This line can actually be just moved out from the for loop. I'll push up a commit for this.